### ISSUE TITLE
Can a component have it's own sub components + viewports?

### ISSUE THREAD

**Issue by PascalPrecht** [angular/router#117]
Giving the fact that each component is somehow isolated and independent, I wonder if it's possible to build a component that introduces sub components and viewports for itself, without coming from an outer routing.

Here's what I mean: We know that we can configure the router to map a route to more than one component like this:

```js
app.controller('AppController', function ($router) {
  $router.config([
    {
      path: '/',
      components: {
        viewportA: 'navigation',
        viewportB: 'main'
      }
    }
  ]);
});
```

Now, when routing to `/` the components `navigation` and `main` are loaded into the viewports `viewportA` and `viewportB`.

Now imagine the following scenario:

We have our `AppController` that maps a route to a `componentA` like this:

```js
app.controller('AppController', function ($router) {
  $router.config([
    {
      path: '/',
      component: 'componentA'
    }
  ]);
});
```

Now, the interesting part, `componentA` has a template like this:

```html
<h1>Component A</h1>
<router-view-port="left"></router-view-port>
<router-view-port="right"></router-view-port>
```

And a router configuration like this:

```js
app.controller('ComponentAController', function ($router) {
  $router.config([
    {
      path: ???
      components: {
        left: 'subComponentA',
        right: 'subComponentB'
      }
    }
  ]);
});
```

So `componentA` is a component, as we know, and it as it's own controller and introduces it's own viewports with it's own router. But as we can see in the code snippet, it's not very clear to what route that maps.

My question is, is that possible at all? That a component is so self-contained that it truly introduces it's own viewports and router? Or do I always have to define my custom "wrapper" route at application level that says "Hey I have this route here and we want to load two components in it.". But that also means, the start template (`index.html`) **needs** to have two viewports.

I hope my question is clear, if not please let me know!

**Comment by jvandemo** [comment]
I think "self-contained" components are essential to component driven Angular development, so yes, I believe it is an absolute must that components can contain their own viewports and have their own router.

If we think of routing as "mounting" components to "paths" where paths are concatenated to form a complete path, then a powerful pattern emerges.

This implies that each component starts from its own root path '/', which is the default path.

So to use your exact example:

```javascript
app.controller('AppController', function ($router) {
  $router.config([
    {
      path: '/',
      component: 'componentA'
    }
  ]);
});
```

and:

```xml
<h1>Component A</h1>
<router-view-port="left"></router-view-port>
<router-view-port="right"></router-view-port>
```

with:

```javascript
app.controller('ComponentAController', function ($router) {
  $router.config([
    {
      path: '/',
      components: {
        left: 'subComponentA',
        right: 'subComponentB'
      }
    }
  ]);
});
```

would concatenate `/` and `/` to form `//` and finally *collapse*  to `/`.

Since each viewport creates a new child router, the paths don't conflict and the current router implementation should already support this.

All kinds of combinations are possible:

`/` + `/` => `/`
`/` + `/sub` => `/sub`
`/main` + `/` => `/main/`
`/main` + `/sub` => `/main/sub`

This is similar to how express works when mounting express apps using a prefix path.

So in short you can just use `/` as the path in your component if you don't want the component route to have its own url (you could consider it the default 'state' of the component when the parent component is is loaded).

Hope that makes sense?

@btford, @PascalPrecht &mdash; would be great to hear your thoughts on this.

**Comment by maximilianmikus** [comment]
+1 express style route mounting for components

**Comment by csaloio1** [comment]
@jvandemo I agree that would be ideal, however I don't think it currently works. Whenever I've tried to get a sub component to map a path such as /sub1 ng-link creates hrefs like /.undefined/sub1

It would be great if it worked as you describe but I don't think it does at least right now. I've yet to see any example of how child viewports/routing can be made to work so far.

**Comment by PascalPrecht** [comment]
@jvandemo Thank you so much for putting this together. At a first glance, this approach makes totes sense. I wonder if there are some edge cases where this idea won't fit though.

Am I getting that proposal right when thinking that this would also solve nested dependencies? Because in your proposal, you'd in fact assemble components via paths.

**Comment by jvandemo** [comment]
> I agree that would be ideal, however I don't think it currently works. Whenever I've tried to get a sub component to map a path such as /sub1 ng-link creates hrefs like /.undefined/sub1

@csaloio1 &mdash; I have noticed that as well and I think it is a bug in the current implementation of the `generate()` method in the router.

**Comment by jvandemo** [comment]
@PascalPrecht &mdash; I think that makes sense as long as a component is *activated* first by its parent component.

So in the example above: even if there was a `componentB` with a `ComponentBController` that matches the same path (`/`) as `componentA`:

```javascript
app.controller('ComponentBController', function ($router) {
  $router.config([
    {
      path: '/',
      components: {
        left: 'subComponentC',
        right: 'subComponentD'
      }
    }
  ]);
});
```

`componentB` would never be loaded, because it wasn't activated by the parent component:

```javascript
app.controller('AppController', function ($router) {
  $router.config([
    {
      path: '/',
      component: 'componentA'
    }
  ]);
});
```

Does that make sense?

Thanks for sharing your thoughts!

**Comment by PascalPrecht** [comment]
@jvandemo Yes, get it!

**Comment by afitterling** [comment]
+1 (btw, has anyone setup a working example on github?)

**Comment by afitterling** [comment]
When requesting the component in main app via CDN from another app leads as SOP says to a view/partal url in the main app based on the domain  requesting the component (the main app). How could I include a foreign component within our app to allow the view partials also to be shared and how to overcome CORS easiliy?

**Comment by csaloio1** [comment]
Does this being on the backlog milestone mean that this is not happening any time soon? Because if so, I can't see one could ever use this router in anything beyond demoware.

**Comment by btford** [comment]
@csaloio1 – this is already implemented. All that remains is to document it better to resolve @PascalPrecht's question.

**Comment by csaloio1** [comment]
@btford That is great to hear. Looking forward to the documentation.

**Comment by upta** [comment]
@btford Just to clarify, does this only work if the child routes have some path beyond just '/'?

In my testing (granted, I'm very new to the new router ;), using the example from @PascalPrecht doing this threw an error:

```
app.controller('ComponentAController', function ($router) {
  $router.config([
    {
      path: "/",
      components: {
        left: 'subComponentA',
        right: 'subComponentB'
      }
    }
  ]);
});
```

but this worked (not allowing for a default setup for child components, since I'd have to route to /test):

```
app.controller('ComponentAController', function ($router) {
  $router.config([
    {
      path: "/test",
      components: {
        left: 'subComponentA',
        right: 'subComponentB'
      }
    }
  ]);
});
```

**Comment by shanonvl** [comment]
@btford that's fantastic that it's implemented but I can't seem to get it to work...  even using a very simple example.   This is using angular `1.4.0.rc.0` (w aria + animate) and angular-new-router `0.5.3`.

http://plnkr.co/edit/hNK8HNjLnGJdPw1APPDM?p=preview

If you can point in the right direction, I'd be happy to help work out some better examples for the documentation.

**Comment by csaloio1** [comment]
@shanonvl Or do what I did and just switch to Aurelia :)

**Comment by AlmeroSteyn** [comment]
@shanonvl I am having the same issue. 

Having `parentComponent` and `childComponent` I set up the following scenario:

Main app controller routing:
```Javascript
 $router.config(
            [
                {
                    path: '/parent',
                    component: 'parentComponent'
                }
            ]
```

This is then loaded into the main html with:

```HTML
<ng-viewport></ng-viewport>
```

Then in the controller of `parentComponent` the following route is defined:

```Javascript
 $router.config(
            [
                {
                    path: '/',
                    component: 'childComponent'
                }
            ]
```

With the html in the view for `parentComponent` containing:

```HTML
<ng-viewport="childComponent"></ng-viewport>
```
This simply does not work. When running it, I do see that the actual html element in the browser is:

```HTML
<ng-viewport="childcomponent"></ng-viewport="childcomponent">
```
So it loses the camelcasing. Dunno if this is why it is losing the connection with the actual component to instantiate.

If I change the routing in the `parentComponent` to:

```Javascript
 $router.config(
            [
                {
                    path: '/child',
                    component: 'childComponent'
                }
            ]
```

And the html of the `parentController` to be the same as the viewport instance on the main page hosting the application:

```HTML
<ng-viewport></ng-viewport>
```

Then I get to see both components on the page when I fist navigate to `/parent` and THEN add on `/child` manually.

Looking forward to getting this working :)

**Comment by PascalPrecht** [comment]
In fact I can't get it working either. Going to investigate now and see what I can do..

**Comment by jvandemo** [comment]
Here is a link to a codepen where sibling viewports are throwing an error: http://codepen.io/jvandemo/pen/WvQeYa?editors=101

![error](https://cloud.githubusercontent.com/assets/1859381/7455225/7f3281ba-f27a-11e4-8e70-9f7f95f5c67d.png)

**Comment by jvandemo** [comment]
I also created a plnkr for easier troubleshooting. It looks like the `path` property is missing from the instructions.

http://plnkr.co/edit/HBPUHhwnzhb9OO461MLI?p=info

![plnkr](https://cloud.githubusercontent.com/assets/1859381/7456277/a3a165bc-f283-11e4-94c4-2f99fafa369c.png)

**Comment by jvandemo** [comment]
@PascalPrecht &mdash; I think a solution is in the making:

https://github.com/angular/angular/commit/9d5c33f9dd66f08e372d1ec051f61f7526562515

**Comment by grbush** [comment]
@jvandemo Great news, my team was wondering about the state of this project, but it looks like there's some very active development over at angular/angular. I wonder if those changes will be getting copied over to angular/router.

**Comment by PascalPrecht** [comment]
Yes. The router code is moved to angular/angular only temporarily in order
to take advantage of the build process. So at some point the latest and
greatest will be moved back to this repo

**Comment by leschirmeur** [comment]
@PascalPrecht any indication when that will happen? We are eagerly waiting for some bug fixes, otherwise we can't use the new router with 1.x. We are currently thinking of switching to ui-router ad interim and then to the new router as soon as fixes are backported to 1.x.

**Comment by PascalPrecht** [comment]
Using ui-router for now should be the right choice when it comes to super
complex scenarios

**Comment by leschirmeur** [comment]
@PascalPrecht Well, super-complex is probably not the proper wording, it's actually quite basic. See e.g. #193

**Comment by jvandemo** [comment]
@leschirmeur &mdash; There has been a lot of activity lately so I think @btford is really working hard to get Component Router ready for Angular 1.4 or Angular 1.5.

The general purpose of making Component Router available in Angular 1.x is to provide a router that allows you to write code that will be easy to port to Angular 2 (once it comes out).

If you need a stable router for your Angular 1.x application today and you don't care about Angular 2 yet, UI-router is probably your best choice.

**Comment by jvandemo** [comment]
@grbush &mdash; @PascalPrecht is right. @btford has mentioned recently that all development is temporarily happening in the `angular/angular` repository.

Once the build process has been finalized, development will continue in the `angular/router` repository.

You can check out Brian's own words right here at the 3 minute mark in:
https://www.youtube.com/watch?v=sfV0ApyN5fM

**Comment by leschirmeur** [comment]
@jvandemo Thank you, we will switch to ui-router then

**Comment by grbush** [comment]
@PascalPrecht @jvandemo  Really appreciate the info thank you. We're super excited about the Component Router, architecting reusable components and basically just planning ahead so we can quickly move to angular2 when it's ready. For now I've cobbled together some 'fixes' (hacks?) that will allow us to prototype with what's in angular / router. Our project schedule allows us time to ship with angular 1.4 and component router 1.0 so eagerly awaiting updates :)

**Comment by jvandemo** [comment]
@leschirmeur If your primary goal is to build a solid Angular 1.x application at the present without thinking about Angular 2 yet, I would make the same call.

**Comment by jvandemo** [comment]
@grbush &mdash; That's a luxury situation ;-) Would love to see your code when you have things running. Is it an open source project? Thanks!

**Comment by leschirmeur** [comment]
@jvandemo -- We have decided to go exactly this route, using the ui-router for the time being. Thanks for your input.

**Comment by AlmeroSteyn** [comment]
@PascalPrecht @jvandemo Thanks for the info! Great to see the direction this is going in. Having the luxury to wait as well. Want to incorporate this is in some Angular training at the company I work for to focus a bit on the upgrade path to 2.0 for future projects.

**Comment by jvandemo** [comment]
@AlmeroSteyn &mdash; That is awesome. Would be great if you could share your experiences. From what we are able to see so far, I believe that Angular 2 and Component Router will be a very powerful combo in the future! Exciting stuff ahead!

**Comment by AlmeroSteyn** [comment]
@jvandemo - I agree. One of the best parts, for me, is how it makes the upgrade to 2.0 less painful. So would want to push for it in new projects the minute it is in stable release. 

Was translating ui-router functionality to new router functionality when I stumbled across this issue.

**Comment by guyjacks** [comment]
Should child routes be working in 0.5.3 or have the changes in angular/angular not been ported over yet?

**Comment by wojciak** [comment]
Bumping this, is there a way to get child components running?

**Comment by jvandemo** [comment]
Here is a link to @btford's talk on the router a few days ago. That should hopefully clear up a lot of the confusion:

https://www.youtube.com/watch?v=z1NB-HG0ZH4
